CRIU throws UnsupportedOperationException when UnsatisfiedLinkError

JasonFengJ9 opened the pull request:
CRIU throws `UnsupportedOperationException` when `UnsatisfiedLinkError` occurred.

`isCRIUSupportEnabled()` returns `false` if the `j9criu29` library wasn't load successfully;
Set `errorMsg` to be returned via `CRIUSupport.getErrorMessage()`.

closes https://github.com/eclipse-openj9/openj9/issues/15554

Signed-off-by: Jason Feng <fengj@ca.ibm.com>

tajila commented:
Rather than this approach we need a way to return false in `isCRIUSupportEnabled` if there is a problem loading the library, and we need `getErrorMessage` to reflect that message

JasonFengJ9 commented:
> Rather than this approach we need a way to return false in isCRIUSupportEnabled if there is a problem loading the library, and we need getErrorMessage to reflect that message

`CRIUSupport.isCRIUSupportEnabled()` is to check if `-XX:+EnableCRIUSupport` was specified for JVM. We moved `System.loadLibrary("j9criu29")` after `InternalCRIUSupport.isCheckpointAllowed()` to improve startup performance.
The `UnsatisfiedLinkError` only appears when loading the library or performing checkpoint, added a field `errorMsg` to record those errors.
Please have another look.

tajila commented:
If a user does `CRIUSupport.isCRIUSupportEnabled()` it would be odd to return an exception that says `UnsupportedOperationException("CRIU support is not enabled")`

In terms of performance, if we called loadNativeLibrary in `CRIUSupport::isCRIUSupportEnabled` it shouldnt affect startup too much since the JCL is using `InternalCRIUSupport::isCRIUSupportEnabled` which doesnt do a library load.

JasonFengJ9 commented:
> If a user does CRIUSupport.isCRIUSupportEnabled() it would be odd to return an exception that says UnsupportedOperationException("CRIU support is not enabled")

It returns `false` and set error message like `CRIU is not enabled - ...`, no exception thrown in this method.

> In terms of performance, if we called loadNativeLibrary in CRIUSupport::isCRIUSupportEnabled it shouldnt affect startup too much since the JCL is using InternalCRIUSupport::isCRIUSupportEnabled which doesnt do a library load.

Moved `init()` into `CRIUSupport::isCRIUSupportEnabled` which returns false if there is a problem loading `j9criu29` library.

JasonFengJ9 commented on jcl/src/openj9.criu/share/classes/org/eclipse/openj9/criu/CRIUSupport.java:
```diff
@@ -176,18 +179,21 @@ public static boolean isCheckpointAllowed() {
+					errorMsg = "There was a problem loaded the criu native library.\n" //$NON-NLS-1$
+							+ "Please check that criu is installed on the machine by running `criu check`.\n" //$NON-NLS-1$
+							+ "Also, please ensure that the JDK is criu enabled by contacting your JDK provider."; //$NON-NLS-1$
+				} else {
```
Removed.

JasonFengJ9 commented on jcl/src/openj9.criu/share/classes/org/eclipse/openj9/criu/CRIUSupport.java:
```diff
@@ -585,24 +591,28 @@ private static RestoreException throwSetEnvException(Throwable cause) {
-			if (InternalCRIUSupport.isCRIUSupportEnabled()) {
-				throw new UnsupportedOperationException(
-					"Running in non-portable mode (only one checkpoint is allowed), and we have already checkpointed once"); //$NON-NLS-1$
+			if (InternalCRIUSupport.isCheckpointAllowed()) {
```
Moved those hooks after `isCheckpointAllowed()`.

tajila commented on jcl/src/openj9.criu/share/classes/org/eclipse/openj9/criu/CRIUSupport.java:
```diff
@@ -585,24 +590,27 @@ private static RestoreException throwSetEnvException(Throwable cause) {
+				try {
+					checkpointJVMImpl(imageDir, leaveRunning, shellJob, extUnixSupport, logLevel, logFile, fileLocks,
+							workDir, tcpEstablished, autoDedup, trackMemory, unprivileged);
+				} catch (UnsatisfiedLinkError ule) {
```
is this exception still possible if we've loaded the library?

tajila replied:
perhaps we should throw an internalVMError instead. I think the message should say that there is something wrong with the JDK

tajila replied:
okay, if we load the library but fail to bind the method then the problem is in the JDK, so I think internalVMError (or something like that) would be more appropriate. The message should indicate that there is a problem with libj9criu in the JDK.

JasonFengJ9 replied:
Missed this notification, `InternalError("There is a problem with libj9criu in the JDK")` is added.

tajila commented:
Jenkins compile xlinuxcriu jdk11
